r1cs/gadgets/crypto/src/prf/ripemd160.rs

```diff
@@ -0,0 +1,458 @@
+            UInt32::addmany(
+                cs.ns(|| format!("compute first T {}", i)),
+                &[result, e.clone()]
+            )
+        }?;
+
+        a = e;
+        e = d;
+        d = c.rotl(10);
+        c = b;
+        b = t;
+
+        t = {
+            let f = apply_round_function(
+                cs.ns(|| format!("second round function {}", i)),
+                &b_prime,
+                &c_prime,
+                &d_prime,
+                79 - i
+            )?;
+            let selected_input_word = x[R_PRIME[i]].clone();
+            let result = UInt32::addmany(
+                cs.ns(|| format!("second rotl(a + f + x + k) {}", i)),
+                &[a_prime, f, selected_input_word, get_round_constants(i).1]
+            )?.rotl(S_PRIME[i]);
+            UInt32::addmany(
+                cs.ns(|| format!("compute second T {}", i)),
+                &[result, e_prime.clone()]
+            )
+        }?;
+
+        a_prime = e_prime;
+        e_prime = d_prime;
+        d_prime = c_prime.rotl(10);
+        c_prime = b_prime;
+        b_prime = t;
+    }
+
+    // Final addition round
+
+    let h0 = UInt32::addmany(
+        cs.ns(|| "compute final t"),
+        &[current_hash_value[1].clone(), c, d_prime]
+    )?;
+
+    let h1 = UInt32::addmany(
+        cs.ns(|| "new h1"),
+        &[current_hash_value[2].clone(), d, e_prime],
+    )?;
+
+    let h2 = UInt32::addmany(
+        cs.ns(|| "new h2"),
+        &[current_hash_value[3].clone(), e, a_prime],
+    )?;
+
+    let h3 = UInt32::addmany(
+        cs.ns(|| "new h3"),
+        &[current_hash_value[4].clone(), a, b_prime],
+    )?;
+
+    let h4 = UInt32::addmany(
+        cs.ns(|| "new h4"),
+        &[current_hash_value[0].clone(), b, c_prime],
+    )?;
+
+    Ok(vec![h0, h1, h2, h3, h4])
+}
+
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use r1cs_std::{alloc::AllocGadget, boolean::AllocatedBit, test_constraint_system::TestConstraintSystem};
+    use algebra::fields::bls12_381::Fr;
+
+    use rand::{RngCore, SeedableRng};
+    use rand_xorshift::XorShiftRng;
+
+    #[test]
+    fn test_full_block() {
+        let mut rng = XorShiftRng::from_seed([
+            0x59, 0x62, 0xbe, 0x3d, 0x76, 0x3d, 0x31, 0x8d, 0x17, 0xdb, 0x37, 0x32, 0x54, 0x06,
+            0xbc, 0xe5,
+        ]);
+
+        let iv = get_ripemd160_iv();
+
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let input_bits: Vec<_> = (0..512)
+            .map(|i| {
+                Boolean::from(
+                    AllocatedBit::alloc(
+                        cs.ns(|| format!("input bit {}", i)),
+                        || Ok(rng.next_u32() % 2 != 0),
+                    )
+                        .unwrap(),
+                )
+            })
+            .collect();
+
+        ripemd160_compression_function(cs.ns(|| "ripemd160"), &input_bits, &iv).unwrap();
+
+        assert!(cs.is_satisfied());
+        assert_eq!(cs.num_constraints() - 512, 18797);
```

Comment:
Was it satisfied? It is a bit less than the number which I predicted in the Google Doc.

Reply:
Everything is correct. The number is slightly lower because of the usage of the MultiEq CS

Reply:
Maybe we can do slightly better using your optimization. I will put a TODO in the code but I don't think it's the correct time to do it